design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+1. A new Rook software version is released, e.g. `v0.6.1`.
+1. The cluster administrator updates their Rook Cluster CRD to update the `versionTag` field using a command similar to `kubectl -n rook edit cluster.rook.io rook`
+1. The Rook operator receives the `update` event for the Cluster CRD instance and sees that the `versionTag` field has changed from its previous value.
+1. The upgrade controller begins a reconciliation to bring the clusters actual version value in agreement with the desired version from the cluster CRD.
+1. **Operator:** The operator pod itself is upgraded first since it is the host of the upgrade controller.
+If there is any new upgrade logic or any migration needed, the new version of the upgrade controller would know how to perform it, so it needs to be updated first.
+The controller uses the Kubernetes API to update the image field of the operator's pod template, which will then begin the process of the deployment that manages the operator pod to terminate the pod and start a new one running the new version in its place.
+    1. When the operator pod comes back up, it needs to remember that it is in an active upgrade process that needs to be resumed.
+    1. The upgrade controller will verify that the operator pod is on the new version and is in a healthy running state.
+1. **API:** The API pod will be upgraded next by updating the `image` field of the pod template spec.
+The `ROOK_VERSION_TAG` env var will also be updated within the template spec.
+The deployment managing the API pod will then terminate the old pod and then start a pod on the new version to replace it.
+    1. The controller will verify that the new API pod is in the `Running` state with the new version and that basic routes are accessible such as `/status`.
+1. **Mons:** The monitor pods will be upgraded in a rolling fashion next.  **For each** monitor, the following actions will be performed by the upgrade controller:
+    1. The `image` field of the pod template spec will be updated to the new version number.
+    Then the pod will be terminated, allowing the replica set that is managing it to bring up a new pod on the new version to replace it.
+    1.  The controller will verify that the new pod is on the new version, in the `Running` state, and that the monitor returns to `in quorum` and has a Ceph status of `OK`.
+    The cluster health will be verified as a whole before moving to the next monitor.
+1. **OSDs:** The OSD pods will be upgraded in a rolling fashion after the monitors.  **For each** OSD, the following actions will take place:
+    1. The `image` field of the pod template spec will be updated to the new version number.
+    1. The lifecycle management of OSDs can be done either as a whole by a single daemon set or individually by a replica set per OSD.
```

Comment:
does the daemonset support rolling updates now? seems like that was coming soon

Reply:
yes, it looks like daemonsets do support rolling updates now, as long as the strategy is set to that (the default is on delete, meaning manually deleting each pod is needed): https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy

i will need to think more about if we should use rolling update, or keep a finer grain control by using "on delete" and doing each pod manually, making sure cluster health overall is OK before moving on. 

I have a feeling that we wouldn't be able to encode full cluster health checking into the rolling update that k8s does, so we may want the finer grain control of "on delete".